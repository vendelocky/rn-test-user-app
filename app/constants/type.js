const type = {
  LOGIN: "login",
  REGISTER: "register",
  ISLOGGEDIN: "isLoggedin",
  ISNOTLOGGEDIN: "isNotLoggedIn",
};

export default type;
