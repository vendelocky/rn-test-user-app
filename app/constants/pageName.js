const pageName = {
    authenticator: "Authenticator",
    userList: "UserList",
};

export default pageName;
