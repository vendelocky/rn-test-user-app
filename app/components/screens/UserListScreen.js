import React, {
  useLayoutEffect,
  useEffect,
  useState,
  useCallback,
  useRef,
} from "react";
import {
  View,
  Image,
  FlatList,
  Text,
  StyleSheet,
  TextInput,
  TouchableOpacity,
  ActivityIndicator,
  BackHandler,
} from "react-native";
import Spacer from "../Spacer";
import { useNavigation } from "@react-navigation/core";
import text from "../../constants/text";
import { useFocusEffect } from "@react-navigation/native";
import StringifyAsyncStorage from "../StringifyAsyncStorage";
import storageKey from "../../constants/storageKey";
import { useSelector } from "react-redux";
import exitAlert from "../exitAlert";
import type from "../../constants/type";
import pageName from "../../constants/pageName";
import colors from "../../constants/colors";

const UserListScreen = () => {
  const navigation = useNavigation();
  const [data, setData] = useState([]);
  const [masterData, setMasterData] = useState([]);
  const [page, setPage] = useState(1);
  const [shouldFetch, setShouldFetch] = useState(true);
  const [stopFetch, setStopFetch] = useState(false);
  const [inputText, setInputText] = useState("");
  const [refreshing, setRefreshing] = useState(false);
  const searchInputText = useRef();
  const selector = useSelector(state => state);

  useFocusEffect(
    React.useCallback(() => {
      const onBackPress = () => {
        if (selector?.Reducer?.status === type.ISLOGGEDIN) {
          exitAlert();
        }
        return true;
      };

      BackHandler.addEventListener("hardwareBackPress", onBackPress);

      return () =>
        BackHandler.removeEventListener("hardwareBackPress", onBackPress);
    }, [selector?.Reducer?.status])
  );

  useLayoutEffect(() => {
    const logOut = () => {
      StringifyAsyncStorage.removeItem(storageKey.TOKEN);
      navigation.canGoBack()
        ? navigation.goBack()
        : navigation.navigate(pageName.authenticator);
    };

    navigation.setOptions({
      headerRight: () => (
        <TouchableOpacity onPress={logOut}>
          <Text style={styles.headerButton}>{text.logout}</Text>
        </TouchableOpacity>
      ),
    });
  }, [navigation]);

  useEffect(() => {
    if (!shouldFetch || stopFetch) {
      return;
    }

    const fetchData = async () => {
      await fetch(`https://reqres.in/api/users?page=${page}`, {
        method: "GET",
      })
        .then((response) => response.json())
        .then((responseJson) => {
          setRefreshing(false);
          setShouldFetch(false);
          setData((prevData) => [...prevData, ...responseJson.data]);
          setMasterData((prevData) => [...prevData, ...responseJson.data]);
          setPage(page + 1);

          if (!responseJson.data.length) {
            setStopFetch(true);
          }
        })
        .catch((error) => {
          setShouldFetch(false);
          console.error(error);
        });
    };

    fetchData();
  }, [page, shouldFetch, stopFetch]);

  const fetchMore = useCallback(() => setShouldFetch(true), []);

  const searchData = (searchedText) => {
    if (searchedText) {
      const newData = masterData.filter((item) => {
        const itemData =
          item.first_name.toUpperCase() + " " + item.last_name.toUpperCase();
        const textData = searchedText.toUpperCase();
        return itemData.indexOf(textData) > -1;
      });
      setData(newData);
    } else {
      setData(masterData);
    }
  };

  const refetch = () => {
    setRefreshing(true);
    setPage(1);
    setShouldFetch(true);
    setStopFetch(false);
    setData([]);
    setMasterData([]);
    setInputText("");
    searchInputText.current.clear();
  };

  const userCard = ({ item }) => {
    return (
      <View style={styles.card}>
        <Image
          source={{ uri: item.avatar }}
          style={styles.image}
          resizeMode={"cover"}
        />
        <View style={styles.itemContainer}>
          <Text style={styles.name}>
            {item.first_name} {item.last_name}
          </Text>
          <Text style={styles.email}>{item.email}</Text>
        </View>
      </View>
    );
  };

  return (
    <View style={styles.root}>
      <View style={styles.searchContainer}>
        <TextInput
          ref={searchInputText}
          style={[styles.SearchBox, styles.searchItems]}
          placeholder={text.searchPlaceholder}
          placeholderTextColor={colors.grey}
          onChangeText={(newText) => setInputText(newText)}
        />
        <TouchableOpacity
          style={[styles.button, styles.searchItems]}
          onPress={() => searchData(inputText)}
        >
          <Text style={styles.searchButton}>{text.searchButton}</Text>
        </TouchableOpacity>
      </View>
      <FlatList
        style={styles.listContainer}
        data={data}
        renderItem={userCard}
        keyExtractor={(item) => item.id.toString()}
        ItemSeparatorComponent={() => <Spacer height={8} />}
        onEndReached={fetchMore}
        onEndReachedThreshold={0.1}
        ListFooterComponent={<Spacer height={24} />}
        onRefresh={refetch}
        refreshing={refreshing}
      />
      {data.length === 0 && !shouldFetch && (
        <View style={styles.empty}>
          <Text style={styles.emptyText}>{text.noResult}</Text>
        </View>
      )}
      {shouldFetch && !stopFetch && (
        <ActivityIndicator size={"large"} color={colors.blue} />
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: colors.white,
  },

  headerButton: {
    color: colors.blue,
    fontSize: 16,
  },

  searchContainer: {
    flexDirection: "row",
    backgroundColor: colors.white,
  },

  searchItems: {
    borderColor: colors.grey,
    borderWidth: 1,
    height: 44,
    margin: 16,
    color: colors.black,
  },

  SearchBox: {
    flex: 1,
    marginRight: 0,
    borderTopLeftRadius: 5,
    borderBottomLeftRadius: 5,
    paddingHorizontal: 16,
  },

  button: {
    marginLeft: 0,
    paddingHorizontal: 24,
    justifyContent: "center",
    borderTopRightRadius: 5,
    borderBottomRightRadius: 5,
  },

  searchButton: {
    color: colors.black,
    fontSize: 16,
    fontWeight: "bold",
  },

  listContainer: {
    padding: 16,
    backgroundColor: colors.white,
  },

  itemContainer: {
    flex: 1,
    paddingVertical: 8,
    borderColor: colors.grey,
    borderBottomWidth: 1,
  },

  card: {
    flexDirection: "row",
  },

  image: {
    marginRight: 16,
    width: 60,
    height: 60,
    borderColor: colors.grey,
    borderBottomWidth: 1,
    borderRadius: 50,
  },

  name: {
    fontWeight: "bold",
    fontSize: 18,
    color: colors.black,
  },

  email: {
    color: colors.darkGrey,
    paddingVertical: 8,
  },

  empty: {
    marginTop: 74,
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    alignItems: "center",
    justifyContent: "center",
  },

  emptyText: {
    fontSize: 16,
    color: colors.black,
  },
});

export default UserListScreen;
