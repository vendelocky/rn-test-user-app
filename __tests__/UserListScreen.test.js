/**
 * @format
 */

import React from "react";
import { cleanup, render } from "@testing-library/react-native";
import UserListScreen from "../app/components/screens/UserListScreen";
import MockedNavigator from "../app/navigations/MockedNavigator";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import type from "../app/constants/type";
import pageName from "../app/constants/pageName";
import renderer from "react-test-renderer";
import { FlatList } from "react-native";

const mockedNavigation = jest.fn();
jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useFocusEffect: () => mockedNavigation,
    useNavigation: () => ({
      navigate: mockedNavigation,
    }),
  };
});

describe("User List Component", () => {
  afterEach(cleanup);
  const mockStore = configureStore();
  const initialState = { status: type.ISNOTLOGGEDIN, initialPage: pageName.authenticator };
  const store = mockStore(initialState);

  it("should match the snapshot", () => {
    const { toJSON } = render(
      <Provider store={store}>
        <MockedNavigator component={UserListScreen} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly", () => {
    renderer.create(
      <Provider store={store}>
        <MockedNavigator component={UserListScreen} />
      </Provider>,
    );
  });

  const testRenderer = renderer.create(
    <Provider store={store}>
      <MockedNavigator component={UserListScreen} />
    </Provider>,
  );
  const testInstance = testRenderer.root;

  describe("Flat List Component", () => {
    const flatList = testInstance.findAllByType(FlatList);
    it("should be defined", () => {
      expect(flatList).toBeDefined();
    });

    it("has only 1 list", () => {
      expect(flatList).toHaveLength(1);
    });
  });
});
