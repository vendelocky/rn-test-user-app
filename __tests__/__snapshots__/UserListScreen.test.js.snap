// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User List Component should match the snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      forwardedRef={[Function]}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      stackPresentation="push"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        hideShadow={false}
        largeTitleHideShadow={false}
        title="MockedScreen"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      >
        <RNSScreenStackHeaderSubview
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          type="right"
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#4099da",
                  "fontSize": 16,
                }
              }
            >
              Logout
            </Text>
          </View>
        </RNSScreenStackHeaderSubview>
      </RNSScreenStackHeaderConfig>
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "rgb(242, 242, 242)",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#ffffff",
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#ffffff",
                "flexDirection": "row",
              }
            }
          >
            <TextInput
              onChangeText={[Function]}
              placeholder="Search..."
              placeholderTextColor="#cccccc"
              style={
                Array [
                  Object {
                    "borderBottomLeftRadius": 5,
                    "borderTopLeftRadius": 5,
                    "flex": 1,
                    "marginRight": 0,
                    "paddingHorizontal": 16,
                  },
                  Object {
                    "borderColor": "#cccccc",
                    "borderWidth": 1,
                    "color": "#000000",
                    "height": 44,
                    "margin": 16,
                  },
                ]
              }
            />
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "borderBottomRightRadius": 5,
                  "borderColor": "#cccccc",
                  "borderTopRightRadius": 5,
                  "borderWidth": 1,
                  "color": "#000000",
                  "height": 44,
                  "justifyContent": "center",
                  "margin": 16,
                  "marginLeft": 0,
                  "opacity": 1,
                  "paddingHorizontal": 24,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#000000",
                    "fontSize": 16,
                    "fontWeight": "bold",
                  }
                }
              >
                GO!
              </Text>
            </View>
          </View>
          <RCTScrollView
            ItemSeparatorComponent={[Function]}
            ListFooterComponent={
              <Spacer
                height={24}
              />
            }
            data={Array []}
            getItem={[Function]}
            getItemCount={[Function]}
            keyExtractor={[Function]}
            onContentSizeChange={[Function]}
            onEndReached={[Function]}
            onEndReachedThreshold={0.1}
            onLayout={[Function]}
            onMomentumScrollBegin={[Function]}
            onMomentumScrollEnd={[Function]}
            onRefresh={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            refreshControl={
              <RefreshControlMock
                onRefresh={[Function]}
                refreshing={false}
              />
            }
            refreshing={false}
            removeClippedSubviews={false}
            renderItem={[Function]}
            scrollEventThrottle={50}
            stickyHeaderIndices={Array []}
            style={
              Object {
                "backgroundColor": "#ffffff",
                "padding": 16,
              }
            }
            viewabilityConfigCallbackPairs={Array []}
          >
            <RCTRefreshControl />
            <View>
              <View
                onLayout={[Function]}
              >
                <View
                  style={
                    Array [
                      null,
                      Object {
                        "height": 24,
                      },
                    ]
                  }
                />
              </View>
            </View>
          </RCTScrollView>
          <ActivityIndicator
            color="#4099da"
            size="large"
          />
        </View>
      </View>
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
